src/include/migraphx/shape_transform_descriptor.hpp

```diff
@@ -0,0 +1,112 @@
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT.  IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ *
+ */
+#ifndef MIGRAPHX_GUARD_MIGRAPHX_SHAPE_TRANSFORM_DESCRIPTOR_HPP
+#define MIGRAPHX_GUARD_MIGRAPHX_SHAPE_TRANSFORM_DESCRIPTOR_HPP
+
+#include <migraphx/config.hpp>
+#include <migraphx/optional.hpp>
+#include <cstdint>
+#include <vector>
+
+namespace migraphx {
+inline namespace MIGRAPHX_INLINE_NS {
+
+struct operation;
+
+// The shape_transform_descriptor class is data structure to simplify shape
+// transformations like reshape, transpose, broadcast, etc. This is made up
+// of a collection of dimensions which are a collection of subdimensions.
+//
+// Each subdimension has an axis and a `len`. The `len` is the length of the
+// subdimension. The axis represents the axis the dimension originated. It is
+// represented as a vector, the first element represents the axis in the
+// original dimension and the additional elements are used when such
+// dimension is split. The axis is empty when its a broadcasted dimension,
+// and a hidden axis can be set if the dimension is associated with a `1`
+// dimension in the original shape.
+//
+// This will first record shape transformations with the `apply` method. This
+// will manipulate this data structure to represent how the transformation
+// changes the dimensions.
+//
+// For example, if we start with an initial dimensions as `[x, y, z]` then
+// each dimension will have one subdimension that corresponds to each
+// original dimension: `[[x:0]], [[y:1]], [[z:2]]`.
+//
+// When a transpose is applied we would just permutate the dimensions.
+//
+// When a reshape that would merge dimensions together then the subdimensions
+// are copied to the same subdimension. So if we reshape the dimensions as `
+// [x*y, z]` then it would become `[[x:0], [y:1]], [[z:2]]`. If the reshape
+// splits the dimension then the subdimension is copied to each dimension and
+// the axis is updated to maintain the order. So a reshape of `[2, x/2, y,
+// z]` would become: `[[2:0,0]], [[x/2:0,1]], [[y:1]], [[z:2]]`.
+//
+// After recording the operators, `simplify` method is used to simplify the
+// data structure such as merging adjacent dimension, etc. The `generate`
+// method is called to generate the operators need to do this
+// transformation.
+struct shape_transform_descriptor
```

Comment:
Many of my questions are stemming from confusion if this `shape_transform_descriptor` is meant to represent only ONE operation or series of operation ?

Reply:
Its meant to represent a series of operations.